### PR TITLE
Document overrides of `clone_from()` in core/std

### PR DESCRIPTION
**Pull request by coolreader18**
As mentioned in https://github.com/rust-lang/rust/pull/96979#discussion_r1379502413

Specifically, when an override doesn't just forward to an inner type, document the behavior and that it's preferred over simply assigning a clone of source. Also, change instances where the second parameter is "other" to "source".

I reused some of the wording over and over for similar impls, but I'm not sure that the wording is actually *good*. Would appreciate feedback about that.

Also, now some of these seem to provide pretty specific guarantees about behavior (e.g. will reuse the exact same allocation iff the len is the same), but I was basing it off of the docs for [`Box::clone_from`](https://doc.rust-lang.org/1.75.0/std/boxed/struct.Box.html#method.clone_from-1) - I'm not sure if providing those strong guarantees is actually good or not.

### CODE REVIEW

**Comment by joboet** [comment]
r? libs-api

File: library/alloc/src/collections/vec_deque/mod.rs [hunk 1]
```diff
@@ -113,9 +113,15 @@ impl<T: Clone, A: Allocator + Clone> Clone for VecDeque<T, A> {
         deq
     }
 
-    fn clone_from(&mut self, other: &Self) {
+    /// Overwrites the contents of `self` with a clone of the contents of `source`.
+    ///
+    /// This method is preferred over simply assigning `source.clone()` to `self`,
+    /// as it avoids reallocation if possible. Additionally, if the element type
+    /// `T` overrides `clone_from()`, this will reuse the resources of `self`'s
+    /// elements as well.
```
Comment (dtolnay) [inline library/alloc/src/collections/vec_deque/mod.rs:121]: It doesn't do this currently. All existing elements are dropped.

**Review by dtolnay — CHANGES_REQUESTED** [review]
Thank you!

**Comment by dtolnay** [comment]
@rust-lang/libs-api:

Please take a skim through the added documentation and decide whether these are all commitments we can make.

Some of them are in the form of _"This method is preferred over simply assigning `source.clone()` to `self`, as it avoids reallocation if possible"_, and it's debatable to what extent that constitutes a commitment. It sounds like something that can be changed saying "yeah we decided it wasn't possible after all".

Other ones are stronger commitments.

Deliberate new commitments:

- `Box::<[T], A>::clone_from` (equivalent to the already existing commitment on `Box::<T, A>::clone_from` for `T: Sized`)
- `LinkedList::<T, A>::clone_from`
- `Vec::<T, A>::clone_from`
- `Waker::clone_from` (https://github.com/rust-lang/rust/pull/96979)

"Commitments":

- `BinaryHeap::<T, A>::clone_from`
- `VecDeque::<T, A>::clone_from`
- `String::clone_from`
- `HashSet::<T, S>::clone_from`
- `OsString::clone_from`
- `PathBuf::clone_from`

@rfcbot fcp merge

**Review by workingjubilee — COMMENTED** [review]
It seems to me "avoids allocations if possible" could also be interpreted to sound like a much more affirmative "we will do this if there is _any possible interpretation_ of our data structure that will avoid allocations here". That may prove to be a bit of a peculiar thing to promise, especially when there's nothing stopping you from using `clone_from` for cloning a larger data structure "into" a smaller one, which will _force_ new allocations.

It is possible that we may wish to say "reduces allocations" instead, as reducing the number of allocations by some N can validly be 0.

I am most concerned about this in the case of the types like BinaryHeap, LinkedList, and the Map/Set types, which have a somewhat more ambiguous internal representation, and in some cases are direly under-loved.

File: library/core/src/cell.rs [hunk 2]
```diff
@@ -1276,8 +1276,8 @@ impl<T: Clone> Clone for RefCell<T> {
     /// Panics if `other` is currently mutably borrowed.
```
Comment (riking) [inline library/core/src/cell.rs:1276]: nit: parameter rename missed in the comment.

**Comment by coolreader18** [comment]
I'm unsure of how the FCP interacts with reviews - should I address the nits/review comments while the FCP is active?

**Comment by dtolnay** [comment]
Please address the review comments at your leisure. It doesn't make a difference whether that happens while the FCP is active or afterward, unless some team member's approval is contingent on seeing the resolution of some concern (usually with "@​rfcbot concern").

**Review by dtolnay — APPROVED** [review]
Thank you! This looks good to me.

I think this wording is all right, but I would be interested to evaluate a followup PR if someone wants to address https://github.com/rust-lang/rust/pull/122201#pullrequestreview-1971822376. The reason I think the current wording in the PR is all right is I don't see "avoids reallocation if possible" being interpreted as "does not allocate if there is any possible interpretation of our data structure that could implement this operation without allocating". It instead just means "avoids those allocations which are possible to avoid per this manifestation of the data structure". The more interesting guarantees being committed to are the ones listed in https://github.com/rust-lang/rust/pull/122201#issuecomment-2028877841.

**Pull request merged**